### PR TITLE
Update response if response id already into payload

### PR DESCRIPTION
**Pull request by OpakAlex**
Trello: https://trello.com/c/t4Q8zGwy/397-update-chattermill-response-huginn-agent

### CODE REVIEW

**Comment by rrodrigu3z** [comment]
Hi @OpakAlex I was looking at the Trello card description, and it says that searching using `user_meta` could be possible too (if that's the case, maybe we'll need to add some options to the configuration form). 

That will be needed or search will be only by id?

File: app/models/agents/chattermill_response_agent.rb [hunk 1]
```diff
@@ -170,24 +169,29 @@ def normalize_response_headers(headers)
       }
     end
 
-    def post_url(event = Event.new)
+    def request_url(event = Event.new)
       event_options = interpolated(event.payload)
       protocol = Rails.env.production? ? 'https' : 'http'
       domain = DOMAINS[Rails.env.to_sym]
       host = "#{event_options['organization_subdomain']}.#{domain}"
+      url = "#{protocol}://#{host}#{API_ENDPOINT}"
```
Comment (rrodrigu3z) [inline app/models/agents/chattermill_response_agent.rb]: Maybe just `"#{protocol}://#{host}#{API_ENDPOINT}/#{event.payload.dig('data', 'id')}"` would work.

**Comment by OpakAlex** [comment]
@rrodrigu3z we choose only via id for now. @duboff approved this.

File: app/models/agents/chattermill_response_agent.rb (hunk 1 above)
Comment (OpakAlex) [reply app/models/agents/chattermill_response_agent.rb]: checking...

**Comment by rrodrigu3z** [comment]
@OpakAlex @duboff Ok got it. I Will test it after my lunch then.

**Comment by OpakAlex** [comment]
@rrodrigu3z pushed some fixes.

**Review by rrodrigu3z — COMMENTED** [review]
@OpakAlex looks good and it seems to be working fine if all attributes are present.

I have only a doubt and it will depend of the use case, 
but, if the receiving event doesn't have all (previous) response `segments` or `user_meta`, the update request will overwrite that data instead of enrich the existing one. Is that what we want?

File: app/models/agents/chattermill_response_agent.rb [hunk 2]
```diff
@@ -170,24 +169,28 @@ def normalize_response_headers(headers)
       }
     end
 
-    def post_url(event = Event.new)
+    def request_url(event = Event.new)
       event_options = interpolated(event.payload)
       protocol = Rails.env.production? ? 'https' : 'http'
       domain = DOMAINS[Rails.env.to_sym]
       host = "#{event_options['organization_subdomain']}.#{domain}"
+     "#{protocol}://#{host}#{API_ENDPOINT}/#{event.payload.dig('data', 'id')}"
```
Comment (rrodrigu3z) [inline app/models/agents/chattermill_response_agent.rb]: It seems that a space is missing (indent)

Comment (OpakAlex) [reply app/models/agents/chattermill_response_agent.rb]: yes, will fix

**Comment by OpakAlex** [comment]
@rrodrigu3z i will recheck this. but if i think if we don't have previous we need override. or @duboff ?

**Comment by rrodrigu3z** [comment]
@OpakAlex Maybe if we're updating a response, maybe we should remove blank attributes at https://github.com/chattermill/huginn/blob/4c82b908f7659906fadc001b522540ce85fed37b/app/models/agents/chattermill_response_agent.rb#L120 because some empty attrs like `created_at` will raise errors.

Maybe something like: `outgoing = interpolated.slice(*BASIC_OPTIONS).select { |_, v| v.present?  }` or similar.

**Comment by OpakAlex** [comment]
good idea @rrodrigu3z ! Thanks

**Comment by rrodrigu3z** [comment]
Looks like the same applies to `check` method.

**Comment by duboff** [comment]
Would it help to use `:patch` instead of `:put`?

**Comment by duboff** [comment]
The idea is to override if the data is in conflict but not if huginn has empty data. Most of the time we will just add a new segment / user_meta variable.

**Comment by OpakAlex** [comment]
@rrodrigu3z pushed fixes, check also this: https://github.com/chattermill/cm-backend/pull/531

File: app/models/agents/chattermill_response_agent.rb [hunk 3]
```diff
@@ -170,24 +169,28 @@ def normalize_response_headers(headers)
...
 
     def handle(data, event = Event.new, headers)
-      url = post_url(event)
+      url = request_url(event)
       headers['Content-Type'] = 'application/json; charset=utf-8'
       body = data.to_json
-      response = faraday.run_request(http_method.to_sym, url, body, headers)
+      method = http_method(event)
+      response = faraday.run_request(method, url, body, headers)
       send_slack_notification(response, event) unless response.status == 201
```
Comment (rrodrigu3z) [inline app/models/agents/chattermill_response_agent.rb]: @OpakAlex One last little thing 😅, now that we're updating, status should be `200` or `201`

**Comment by OpakAlex** [comment]
@rrodrigu3z pushed fix

**Comment by rrodrigu3z** [comment]
👍

**Pull request merged**